rfc/rfc-0000-common-blueprint.md

```diff
@@ -0,0 +1,220 @@
+# Meta
+[meta]: #meta
+- Name: Common Blueprint Architecture
+- Start Date: 2022-03-29
+- Author(s): @emmjohnson @martyspiewak
+- Status: Draft
+- RFC Pull Request: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC proposes collapsing all current "Blueprints" and templates (not including `ClusterRunTemplate`) into a common type, `ClusterBlueprint`. 
+As part of this, the RFC proposes that outputs be dynamically typed via a new CR `ClusterBlueprintType` and remove the statically typed outputs that exist today (url, revision, image, config). This allows us to maintain the contract of swappable templates, without restricting outputs to a predefined set. 
+Additionally, the selectors will be removed from blueprints and instead be specified in a new CR `ClusterBlueprintSelector`.
+
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+
+  - Today we have seven CRDs that are needed to create a supply chain and delivery. To simplify the data model for users, we can
+collapse all seven CRDs into a few common CRDs. Additionally, whereas today the overhead to start reconciling templates would be adding five identical
+reconcilers, this would enable us to have one reconciler for all blueprints and more easily add meaningful data to their statuses.
+
+- What use cases does it support?
+
+  - This would allow for snippets and users to specify any output they want, while maintaining strong typing.
+  - The `ClusterBlueprintSelector` will enable users to separate RBAC between template authors and operators who want to restrict what blueprints can be selected.
+
+- What is the expected outcome?
+
+  - Simplified on-boarding and user experience.
+
+# What it is
+[what-it-is]: #what-it-is
+* `ClusterBlueprints` will encapsulate everything the current templates allow, while adding `spec.resources` to enable defining multiple resources (i.e. snippets).
+* A `ClusterBlueprint` can take one of `spec.resources`, `spec.template`, and `spec.ytt`. It will allow inlining of templates in `spec.resources`.
+* A `ClusterBlueprint` must specify a `blueprintTypeRef` which defines the outputs required from the blueprint. The `ClusterBlueprint`'s `spec.outputs` must match those outputs.
+* A `ClusterBlueprintType` defines only `spec.outputs`, which is an array of output names. By leaving the `name` key in the array, we leave open the possibility to later add `type` to each output, though this RFC currently proposes only supporting string outputs.
```

Comment:
> _though this RFC currently proposes only supporting **string** outputs._

in the sense that any `valuePath` that evaluates to anything other than a string wouldn't be considered valid? I think that'd be a big regression as, e.g., those leveraging today `ClusterConfigTemplate` with `configPath` pointing at a map structure like a podtemplatespec would not be able to do so anymore

Reply:
configPath is of type `string`  today - https://github.com/vmware-tanzu/cartographer/blob/120a53cd8577364610563969b27183f83242a4ae/config/crd/bases/carto.run_clusterconfigtemplates.yaml#L58

Reply:
ooooooh I see, I was thinking about the type of the value from the jsonpath query, makes sense 👍